Mempool syncing optimizations & fixups

shesek opened the pull request:
Optimizations:

- Send RPC requests in parallel over multiple TCP connections, without batching (38e73d9a2808586e202b16d619645e671817c055)

  The number of parallel requests to send can be controlled with a new `--daemon-parallelism` CLI options (defaults to 4).

- Continuously attempt to fetch a full mempool snapshot, keeping whatever transactions we were able to get each time (6a14f40d61317b11b6b3f0eeae1199d39585dda9)

Fixups (not strictly necessary for the other changes, but touches on related areas and the previous behavior was wrong):

- Don't add transactions with unknown parents to ensure mempool consistency (c60207ad6c81f1ec05982babf7941be53ddcdb26)

- Make sure the chain tip doesn't move while fetching the mempool (a6a796c69e5ea935ff67434633139f0dd214476e)

This will require adjusting bitcoind's `rpcthreads` and possibly `rpcworkqueue` configuration options upwards. The default `rpcthreads` is 4, which would be exhausted by mempool fetching alone (`electrs` keeps at least one other open connection, and we probably want to allow some room for other clients too).

Based on top of #76. (now merged)

shesek committed:
Don't use RPC batching with bitcoind

This actually hurts performance because the batched response has to be
bueffered on the bitcoind side, as @TheBlueMatt explains at
https://github.com/romanz/electrs/issues/373#issuecomment-785533444

Instead, send multiple individual RPC requests in parallel using a
thread pool, with a separate RPC TCP connection for each thread.

Also see https://github.com/romanz/electrs/pull/374

shesek committed:
Continuously attempt to fetch mempool transactions

Previously, if any of the mempool transactions were not available
because they were evicted between getting the mempool txids and txs
themselves, the mempool syncing operation would be aborted and tried
again from scratch.

With this change, we instead keep whatever transactions we were able to
fetch, then get the updated list of mempool txids and re-try fetching
missing ones continuously until we're able to get a full snapshot.

shesek committed:
Don't store mempool txs before all prevouts are available

The indexing process was adding transactions into the store so that
prevouts funded & spent within the same batch could be looked up via
Mempool::lookup_txos(). If the indexing process later failed for any
reason, these transactions would remain in the store.

With this change, we instead explicitly look for prevouts funded within
the same batch, then look for the rest in the chain/mempool indexes and
fail if any are missing, without keeping the transactions in the store.

shesek committed:
Make sure the chain tip doesn't move while fetching the mempool

philippem commented:
Overall this MR seems to greatly improve the initial sync and steady state of electrs alongside a slower bitcoind.

some issues found during testing
- the electrum method `blockchain.transaction.get` will panic on the handling thread if the transaction is not known by electrs (panic starts at https://github.com/shesek/electrs/blob/202405-mempool-sync/src/electrum/server.rs#L375 
- some requests from electrs to bitcoind will disconnect waiting for recv, at https://github.com/shesek/electrs/blob/202405-mempool-sync/src/daemon.rs#L210
Increasing the `rpcthreads` parameter in bitcoind does not seem to change this. Electrs itself seems to recover. These recv disconnects happen with different rpcs (usually `getrawtransaction`, but also `getbestblockhash`)

RCasatta commented on src/new_index/schema.rs:
```diff
@@ -1038,26 +1038,19 @@ fn lookup_txos(
+    let mut remain_outpoints = outpoints.iter();
+    txstore_db
+        .multi_get(outpoints.iter().map(TxOutRow::key))
+        .into_iter()
```
using `zip(outpoints.iter())` is probably more conventional

shesek replied:
Indeed, thanks! Not sure why I didn't do that. Updated in #76.

RCasatta commented on src/new_index/schema.rs:
```diff
@@ -1038,26 +1038,19 @@ fn lookup_txos(
+        .into_iter()
+        .filter_map(|res| {
+            let outpoint = remain_outpoints.next().unwrap();
+            match res.unwrap() {
```
slightly better to use `expect`?

shesek replied:
Possibly, but the codebase currently mostly uses `unwrap()` for unexpected errors from rocksdb that we don't have a specific cause/explanation to attach to

RCasatta commented on src/daemon.rs:
```diff
@@ -436,13 +427,45 @@ impl Daemon {
     }
 
+    fn retry_reconnect(&self) -> Daemon {
+        // XXX add a max reconnection attempts limit?
```
maybe also a sleep between retries?

shesek replied:
What do you think @philippem?

philippem replied:
yes, seems like a good idea

RCasatta replied:
Do we want to do something like [this](https://github.com/Blockstream/electrs/commit/4a455d49d3a898343ce735773f83feaebed37ceb)? 

I am not sure honestly but I don't want to keep this MR hanging for this

RCasatta commented on src/daemon.rs:
```diff
@@ -436,13 +427,45 @@ impl Daemon {
     fn requests(&self, method: &str, params_list: &[Value]) -> Result<Vec<Value>> {
-        self.retry_request_batch(method, params_list)
+        let thread_pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(self.daemon_parallelism)
```
would be better to be `self.daemon_parallelism.min(params_list.len())` ?

shesek replied:
My understanding is that rayon will internally only initialize as many threads as necessary, so this should not be needed

RCasatta commented on src/daemon.rs:
```diff
@@ -436,13 +427,45 @@ impl Daemon {
+    // Send requests in parallel over multiple connections as individual JSON-RPC requests (with no JSON-RPC batching)
     fn requests(&self, method: &str, params_list: &[Value]) -> Result<Vec<Value>> {
-        self.retry_request_batch(method, params_list)
+        let thread_pool = rayon::ThreadPoolBuilder::new()
```
Are there specific reasons to create a thread pool here instead of using the rayon default one which can be build once with the `daemon_parallelism` arg?

In the past I had issues in libraries using the default thread pool (because if downstream someone use it may break everything) but here is a binary and should be ok.

It's ok if we want to stay on the safe path just wondering if there is some other reasoning behind it.

shesek replied:
Well, for one, I was not aware that reconfiguring the number of threads used by the default rayon pool was possible. :)

But I still don't think that we should. We're setting up multiple rayon thread pools with different parallelism configurations (3 - for bulk block parsing, stats pre-caching, and in this PR), so it doesn't really make sense to use one of the configurations as the default.

Another consideration I had in mind when implementing this was to have the TCP connections terminate when the pool is dropped. But thinking about this now, its seems better to reuse the same (custom) thread pool and keep these connections open, to reduce TCP connection overhead. I implemented this in beef521f9841530fab5e6917047f9e31c07cc8e4.

RCasatta commented on src/daemon.rs:
```diff
@@ -510,12 +533,7 @@ impl Daemon {
-        }
-        assert_eq!(txhashes.len(), txs.len());
-        Ok(txs)
+        values.into_iter().map(tx_from_value).collect()
```
Unrelated to this PR, but I noticed `tx_from_value` (and `block_from_value`) could probably be improved by using rust bitcoin `deserialize_hex` which doesn't allocate the intermediate vector of bytes.
Just noticed the rust bitcoin version we are currently using doesn't have that yet, leaving this here anyway if we want to upgrade in the future.

RCasatta replied:
Referenced as issue, resolving...

RCasatta commented on src/new_index/mempool.rs:
```diff
@@ -291,7 +291,11 @@ impl Mempool {
-                self.add(vec![tx])
+                self.add({
+                    let mut txs_map = HashMap::new();
+                    txs_map.insert(tx.txid(), tx);
```
should reuse the existing txid instead of recomputing it which is expensive

shesek replied:
Updated in a6c181e2f12433860a100c4b44b8dc19818f4afb.

RCasatta commented on src/new_index/mempool.rs:
```diff
@@ -300,17 +304,13 @@ impl Mempool {
     }
 
-    fn add(&mut self, txs: Vec<Transaction>) -> Result<()> {
+    fn add(&mut self, txs_map: HashMap<Txid, Transaction>) -> Result<()> {
```
is this function  called just from `add_by_txid` ?
If so this map always contains one value,  thus I would accept singular parameter txid, tx and then initiate the map internally.

shesek replied:
No, it is also used by `Mempool::update()` with multiple transactions.

RCasatta commented on src/daemon.rs:
```diff
@@ -518,22 +526,37 @@ impl Daemon {
-        let params_list: Vec<Value> = txhashes
+    /// Fetch the given transactions in parallel over multiple threads and RPC connections,
+    /// ignoring any missing ones and returning whatever is available.
+    pub fn gettransactions_available(&self, txids: &[&Txid]) -> Result<Vec<Transaction>> {
```
I would return `(Txid, Transaction)` so that caller can avoid recomputing txid

shesek replied:
Updated in a6c181e2f12433860a100c4b44b8dc19818f4afb.

RCasatta reviewed (dismissed):
utACK a6a796c69e5ea935ff67434633139f0dd214476e code review

My comments are mostly styling and perf improv.

Note also I am not that familiar with the overall codebase, but all changes seems reasonable/

philippem commented:
>     * some requests from electrs to bitcoind will disconnect waiting for recv, at https://github.com/shesek/electrs/blob/202405-mempool-sync/src/daemon.rs#L210
>       Increasing the `rpcthreads` parameter in bitcoind does not seem to change this. Electrs itself seems to recover. These recv disconnects happen with different rpcs (usually `getrawtransaction`, but also `getbestblockhash`)

While tracing this with wireshark, I observed that electrs sends a request (such as getbestblockhash), and waits for a response. After 30 seconds (and a few TCP Keep-Alive interactions), bitcoind sends a FIN,ACK without any text response. Seems to be some behaviour in bitcoind.

shesek committed:
Update logging verbosity

- Reduce logging level for bitcoind's JSONRPC response errors

  These can happen pretty often for missing mempool txs and do not
  warrant warn-level logging. Unexpected RPC errors will bubble up and
  be handled appropriately.

- Add more verbose logging for mempool syncing

shesek committed:
Reuse RPC threads and TCP connections

Keep RPC TCP connections open between sync runs and reuse them,
to reduce TCP connection initialization overhead.

shesek committed:
Avoid recomputing txids when possible

Following https://github.com/Blockstream/electrs/pull/89#discussion_r1632848608
and https://github.com/Blockstream/electrs/pull/89#discussion_r1632919266

shesek commented:
> the electrum method `blockchain.transaction.get` will panic on the handling thread if the transaction is not known by electrs 

I was unable to reproduce this, the "missing transaction" message appears to be returned as a JSONRPC error without panicking:

```
# Client
$ echo '{"id": 1, "method": "blockchain.transaction.get", "params": ["0000000000000000000000000000000000000000000000000000000000000000"]}' | nc 172.17.0.3 60401
{"error":"missing transaction","id":1,"jsonrpc":"2.0"}

# Server logs
TRACE RPC Request("{\"id\": 1, \"method\": \"blockchain.transaction.get\", \"params\": [\"0000000000000000000000000000000000000000000000000000000000000000\"]}\n")
WARN rpc #1 blockchain.transaction.get [String("0000000000000000000000000000000000000000000000000000000000000000")] failed: Error: missing transaction
TRACE RPC Done
# ... keeps running
```

> Seems to be some behaviour in bitcoind.

I'm seeing that behaviour on my local regtest env too.

RCasatta commented:
utACK a6c181e2f12433860a100c4b44b8dc19818f4afb

RCasatta reviewed (dismissed):
utACK f3a13e7c3c626c86c9709952651a2c90a9df0a70

philippem commented:
in analyzing pcap trace between local electrs and local. bitcoin rpc, we see electrs asking for getrawmempool on 
3 times in 16 seconds, in a single socket connection. MR needs more analysis/testing.

philippem commented:
2 issues
- repeated requests for full mempool on one thread (correct?)

- disconnects are actually coming from electrs, not bitcoind (screenshot of pcap)

![image](https://github.com/user-attachments/assets/2033958a-5a49-43e0-8cf4-940f9dd021fd)

philippem commented:
```
2024-08-16T09:26:16.938-07:00 - INFO NetworkInfo { version: 270000, subversion: "/Satoshi:27.0.0/", relayfee: 1e-5 }
2024-08-16T09:26:16.938-07:00 - INFO BlockchainInfo { chain: "main", blocks: 857053, headers: 857053, bestblockhash: "00000000000000000003066eeec0d73da4cf50ee6bf7e31f9032145c17ad3d54", pruned: false, verificationprogress: 0.9999971, initialblockdownload: Some(false) }
2024-08-16T09:26:34.141-07:00 - INFO hash=00000000000000000003066eeec0d73da4cf50ee6bf7e31f9032145c17ad3d54 height=857053 @ 2024-08-16T16:18:23Z (2 left to index)
2024-08-16T09:27:14.773-07:00 - INFO Electrum RPC server running on 0.0.0.0:50001
2024-08-16T09:27:14.777-07:00 - INFO REST server running on 0.0.0.0:3001
2024-08-16T09:27:19.773-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.264-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.264-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.265-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.266-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.267-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:23.267-07:00 - WARN reconnecting to bitcoind: disconnected from daemon while receiving
2024-08-16T09:27:26.691-07:00 - WARN chain tip moved while updating mempool
2024-08-16T09:27:26.691-07:00 - WARN skipped failed mempool update, trying again in 5 seconds
2024-08-16T09:27:31.692-07:00 - INFO hash=000000000000000000008e575ea9a5b9bcd0591f6b749f975fd679fc8ca23e62 height=857054 @ 2024-08-16T16:26:43Z (1 left to index)

```

philippem commented:
following one of the "disconnects":

1. electrs makes RPC request for "getrawtransaction"
2. bitcoind responds with data
3. electrs sits on connection for 30 seconds
4. bitcoin sends FIN,ACK but electrs does not close
5. electrs attempts to send more
6. bitcoin sends RST

electrs should be responding to the FIN,ACK and active, and reopen

![image](https://github.com/user-attachments/assets/4521ecc0-e6bc-43b5-8ff4-aa781f27395c)

philippem reviewed (changes requested):
see earlier comments about connection handling.

philippem commented:
- we are also seeing 5 second delays after receiving 500 from bitcoin. The 500 error is used by bitcoin rpc to indicate that the getrawtransaction call requested a non existent tx, which will happen after blocks are mined, and in presence of full rbf. Electrs should consider this case of error by inspecting the details of the json-rpc response, and not wait 5 seconds.

shesek commented:
> repeated requests for full mempool on one thread ... we see electrs asking for `getrawmempool` on 3 times in 16 seconds, in a single socket connection

`getrawmempool` is used to get the list of mempool txids (not txs), and should be sent from the main (sync loop) thread using the same socket (unless it disconnects). It should be sent at least once every 5 seconds, possibly more if additional roundtrips are needed to get a full mempool view.

> The 500 error is used by bitcoin rpc to indicate that the getrawtransaction call requested a non existent tx ... Electrs should consider this case of error by inspecting the details of the json-rpc response, and not wait 5 seconds.

This should already be the case. In fact only errors explicitly constructed as an `ErrorKind::Connection` would trigger the automatic reconnection behavior, other errors will bubble up. RPC errors for "tx not found" are [identified and ignored](https://github.com/Blockstream/electrs/pull/89/files#diff-43553fa84fd0a0c6bc4d2fd3926dbe8c6019b1691cc3ba441ffb8a33b70eb1f7R552), while other unexpected RPC errors terminate execution (without reconnecting).

> disconnects are actually coming from electrs, not bitcoind (screenshot of pcap)

Looking into this.

shesek commented:
> electrs sits on connection for 30 seconds .. bitcoin sends FIN,ACK but electrs does not close

bitcoind appears to close the connection after a period of inactivity, which can be adjusted with the (hidden) config option `rpcservertimeout` (defaults to 30 seconds).

Adjusting it upwards made the disconnection issues I was experiencing in my local env go away. I also tested with electrs and bitcoind connected over the internet (local electrs connecting to a regtest bitcoind on a vps) and was able to retain open TCP connections with no disconnections for several hours (used every ~2 minutes, with `rpcservertimeout` set to 5 minutes).

I also tried investigating what's causing electrs not to close immediately in reply to bitcoind's FIN, but do not have an answer yet.

shesek commented:
Rebased and added some more verbose logging for mempool syncing.

philippem commented:
testing latest commit using local electrs and bitcoind, I set parameters in bitcoin.conf

```
rpcthreads=16
rpcservertimeout=600
rpcworkqueue=50
```

I have not seen any "disconnected from daemon while receiving" messages.

RCasatta commented:
utACK 7a068bf0e0dc3c063b0f661bc178fd8b4c3e2290

philippem commented:
These are the guidelines for the bitcoin.conf rpc settings:

D = --daemon-parallelism D for electrs, should be # of cores given to container, default is 2

E = # of electrs instances

B = # of bitcoin core instances

Then in bitcoin.conf set

```
rpcthreads=ceil(E/B) * (D + 3)
rpcworkqueue=ceil(E/B) * D * 8
rpcservertimeout=600

```
that is, we provide D + 3 threads for each electrs instead, and D*8 queue for each electrs.
This assumes that only electrs is using bitcoin core.
For example, if we have an 3 electrs each with 4 cores, using 2 bitcoind, then we should have

```
--daemon-parallelism=4  # option to electrs
```

bitcoin.conf settings:
```
rpcthreads=14
rpcworkqueue=64
rpcservertimeout=600
```
